astropy/modeling/powerlaws.py

```diff
@@ -583,28 +583,36 @@ class Schechter1D(Fittable1DModel):
 
     phi_star = Parameter(default=1., description=('Normalization factor '
                                                   'in units of number density'))
-    m_star = Parameter(default=-20., description='Characteristic magnitude')
+    m_star = Parameter(default=-20., description='Characteristic magnitude', mag=True)
     alpha = Parameter(default=-1., description='Faint-end slope')
 
     @staticmethod
-    def evaluate(mag, phi_star, m_star, alpha):
+    def _factor(magnitude, m_star):
+        factor_exp = (magnitude - m_star)
+
+        if isinstance(factor_exp, Quantity):
+            if factor_exp.unit == mag:
+                factor_exp = Magnitude(factor_exp.value, unit=mag)
+
+                return factor_exp.to(dimensionless_unscaled)
+            else:
+                raise UnitsError("The units of magnitude and m_star must be a magnitude")
```

Comment:
> It is slightly worse than that (and I can say that as I wrote the code...), `u.mag` is not the same as `u.mag()` - only the latter is a proper logarithmic unit. In practice, the two behave mostly the same, but one cannot get the auto-conversion to dimensionless from the plain `Quantity`:
> 
> ```
> In [3]: (2.5*u.mag()).to(u.one)
> Out[3]: <Quantity 0.1>
> 
> In [4]: (2.5*u.mag).to(u.one)
> UnitConversionError: 'mag' and '' (dimensionless) are not convertible
> ```
> 
> I should probably note that really this is not a trivial problem: magnitudes (and dB and dex) are rather strange units, which behave differently!

@mhvk does this address the issue? Note that there is are now test cases in `test_model_quantities.py:test_models_evaluate_magunits` with both `u.mag` and `u.ABmag` units.

Reply:
I'm confused how this can work for `ABmag`?! A `Magnitude` would pass the `isinstance` check, but the unit would not match. What am I missing?!

I'd have thought it would be something like this:
```
if isinstance(factor_exp, Quantity):
    try:
        if not isinstance(factor_exp, Magnitude):
            factor_exp = Magnitude(factor_exp.to_value(u.mag))
        return factor_exp.to(dimensionless_unscaled)
    except UnitsError as err:
        raise UnitTypeError("the units of magnitude and m_star must be magnitudes with matching physical type.") from err
```

Reply:
It is operating on `magnitude - m_star`.  Lets look at what gets returned:
```python
import astropy.units as u

magnitude_1 = -23 * u.ABmag
magnitude_2 = -23 * u.mag
m_star_1 = -20 * u.ABmag
m_star_2 = -20 * u.mag

factor_exp_1 = magnitude_1 - m_star_1
factor_exp_2 = magnitude_2 - m_star_2

print(isinstance(factor_exp_1, u.Quantity))
print(isinstance(factor_exp_2, u.Quantity))

print(factor_exp_1.unit == u.mag)
print(factor_exp_2.unit == u.mag)
```

Returns:
```python
True
True
True
True
```

It turns out that the unit on the difference of two `ABmag` quantities is `mag`.

Reply:
Ah, yes, of course! In principle that will change with #12579, which is why I was confused. But I think really your statement should work after that too, so I better just fix it so that `u.mag() == u.mag`. Anyway, no concerns for *this* PR, then!